lectures/sympy.md

```diff
@@ -0,0 +1,861 @@
+
+We can construct a simple series of summations using `Sum` function
+
+```{code-cell} ipython3
+x, y, i, j = symbols("x y i j")
+sum_xy = Sum(Indexed('x',i) * Indexed('y', j), 
+            (i, 0, 3),
+            (j, 0, 3))
+sum_xy
+```
+
++++ {"user_expressions": []}
+
+To evaluate the sum, we can `lamdify` the formula.
+
+The lamdified expression can take vectors as input for $x$ and $y$ and compute the result
+
+```{code-cell} ipython3
+sum_xy = lambdify([x, y], sum_xy)
+grid = np.arange(0, 4, 1)
+sum_xy(grid, grid)
+```
+
++++ {"user_expressions": []}
+
+#### Example: bank deposits
+
+Imagine a bank with $D_0$ as the deposit at time $t$, it loans $(1-r)$ of its deposits and keeps a fraction $r$ as cash reserves.
+
+Its deposits at time is
+
+$$
+\sum_{i=0}^\infty (1-r)^i D_0
+$$
+
+```{code-cell} ipython3
+r, D = symbols('r D')
+
+Dt = Sum('(1 - r)^i * D', (i, 0, oo))
+Dt
+```
+
++++ {"user_expressions": []}
+
+We can call `doit` method to evaluate the series
+
+```{code-cell} ipython3
+Dt.doit()
+```
+
++++ {"user_expressions": []}
+
+Simplifying the expression above gives
+
+```{code-cell} ipython3
+simplify(Dt.doit())
+```
+
++++ {"user_expressions": []}
+
+This is consistent with the solution we provided in our lecture on [geometric series](https://python.quantecon.org/geom_series.html#example-the-money-multiplier-in-fractional-reserve-banking).
+
+
+#### Example: discrete random variable
+
+We can also use SymPy to compute the expectation of a discrete random variable.
+
+Let's define a discrete random variable $X$ with [Poisson distribution](https://en.wikipedia.org/wiki/Poisson_distribution):
+
+$$
+f(x) = \frac{\lambda^x e^{-\lambda}}{x!}, \quad x = 0, 1, 2, \ldots
+$$
+
+```{code-cell} ipython3
+λ = symbols('lambda')
+
+# We refine the symbol x to positive integers
+x = Symbol('x', integer=True, positive=True)
+pmf = λ**x * exp(-λ)/factorial(x)
+pmf
+```
+
+We can verify if the distribution follows the fundamental property of probability distribution:
+
+$$
+\sum_{x=0}^{\infty} f(x) = 1
+$$
+
+```{code-cell} ipython3
+sum_pmf = Sum(pmf, (x, 0, oo))
+sum_pmf.doit()
+```
+
+The expectation of the distribution is:
+
+$$
+E(X) = \sum_{x=0}^{\infty} x f(x) 
+$$
+
+```{code-cell} ipython3
+fx = Sum(x*pmf, (x, 0, oo))
+fx.doit()
+```
+
+```{note}
+
+SymPy offers a statistics module called [`Stats`](https://docs.sympy.org/latest/modules/stats.html).
+
+`Stats` module offers built-in distributions and functions on probability distributions.
+
+The operation above can also be computed using `Stats` module.
+```
+
++++ {"user_expressions": []}
+
+## Symbolic Calculus
+
+SymPy allows us to perform various calculus operations, such as differentiation and integration.
+
+
+### Limits
+
+We can compute limits for a given function using the `limit` function
+
+```{code-cell} ipython3
+---
+mystnb:
+  image:
+    width: 3%
+---
+# Define a function
+f = x ** 2 / (x - 1)
```

Comment:
This is actually an expression and not a function in sympy. Function `F` would be:
```
F = Lambda(x, f)
```